dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/trackedentity/HibernateTrackedEntityStore.java

```diff
@@ -710,28 +710,65 @@ private String getFromSubQueryJoinOrgUnitConditions(TrackedEntityQueryParams par
       return orgUnits.toString();
     }
 
-    if (params.isOrganisationUnitMode(OrganisationUnitSelectionMode.DESCENDANTS)) {
-      SqlHelper orHlp = new SqlHelper(true);
+    String orgUnitSql = getOrgUnitSql(params, "OU");
 
-      orgUnits.append("AND (");
+    if (!Strings.isNullOrEmpty(orgUnitSql)) {
+      orgUnits.append(" AND (").append(orgUnitSql).append(") ");
+    }
 
-      for (OrganisationUnit organisationUnit : params.getAccessibleOrgUnits()) {
+    return orgUnits.toString();
+  }
 
-        OrganisationUnit ou = organisationUnitStore.getByUid(organisationUnit.getUid());
-        if (ou != null) {
-          orgUnits.append(orHlp.or()).append("OU.path LIKE '").append(ou.getPath()).append("%'");
-        }
-      }
+  private String getOrgUnitSql(TrackedEntityQueryParams params, String ouTable) {
+    return switch (params.getOrgUnitMode()) {
+      case SELECTED -> getSelectedOrgUnitPath(params.getAccessibleOrgUnits(), ouTable);
+      case CHILDREN -> getChildrenOrgUnitsPath(params.getAccessibleOrgUnits(), ouTable);
+      case ALL -> null;
+      default -> getOrgUnitsPath(params.getAccessibleOrgUnits(), ouTable);
+    };
+  }
 
-      orgUnits.append(") ");
-    } else if (!params.isOrganisationUnitMode(OrganisationUnitSelectionMode.ALL)) {
-      orgUnits
-          .append("AND OU.organisationunitid IN (")
-          .append(getCommaDelimitedString(getIdentifiers(params.getAccessibleOrgUnits())))
-          .append(") ");
+  private String getChildrenOrgUnitsPath(Set<OrganisationUnit> orgUnits, String ouTable) {
+    StringJoiner orgUnitSqlJoiner = new StringJoiner(" or ");
+
+    for (OrganisationUnit orgUnit : orgUnits) {
+      orgUnitSqlJoiner.add(
+          ouTable
+              + "."
+              + PATH_LIKE
+              + " '"
+              + orgUnit.getPath()
+              + "%' "
+              + " and ("
+              + ouTable
+              + "."
+              + "hierarchylevel = "
+              + orgUnit.getLevel()
+              + " or "
+              + ouTable
+              + "."
+              + "hierarchylevel = "
+              + (orgUnit.getLevel() + 1)
```

Comment:
I see the old code does not use the +1, but I am pretty sure it should onyl be +1; which means it might have been a bug before.